### ISSUE TITLE
stack overflow in out_lib_flush

### ISSUE THREAD

**Issue by donbowman** [fluent/fluent-bit#797]
the bin/flb-rt-in_cpu test will show this.
Compile with -DSANITIZE_ADDRESS=On (from my PR https://github.com/fluent/fluent-bit/pull/792 that adds sanitizers)

If we look on line 131:

```
    while (msgpack_unpack_next(&result, data, bytes, &off)) {
        switch(ctx->format) {
        case FLB_OUT_LIB_FMT_MSGPACK:
            /* copy raw bytes */
            data_for_user = flb_malloc(bytes);
            if (!data_for_user) {
                flb_errno();
                msgpack_unpacked_destroy(&result);
                FLB_OUTPUT_RETURN(FLB_ERROR);
            }
            memcpy(data_for_user, &result.data, bytes); *********
            data_size = bytes;
            break;
```
we are copying from &result.data, which is on the stack, about 11KB.
Maybe it should be result.data.via.array.ptr ?

```
Test flush_2s_2times...                         [2018/09/25 18:01:10] [ info] [engine] started (pid=3558)
=================================================================
==3558==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x62b000006320 at pc 0x7f1de0877733 bp 0x62b000006180 sp 0x62b000005928
READ of size 1105 at 0x62b000006320 thread T1
    #0 0x7f1de0877732  (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x79732)
    #1 0x55cbd17dac81 in out_lib_flush /home/don/src-ag/platform/fluent-bit/plugins/out_lib/out_lib.c:131
    #2 0x55cbd16d5f6f in output_pre_cb_flush /home/don/src-ag/platform/fluent-bit/include/fluent-bit/flb_output.h:311
    #3 0x55cbd19f3e86 in co_init /home/don/src-ag/platform/fluent-bit/lib/monkey/deps/flb_libco/amd64.c:117

0x62b000006400 is located 0 bytes to the right of 25088-byte region [0x62b000000200,0x62b000006400)
allocated by thread T1 here:
    #0 0x7f1de08dcb50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x55cbd19f3fa0 in co_create /home/don/src-ag/platform/fluent-bit/lib/monkey/deps/flb_libco/amd64.c:142
    #2 0x55cbd16d6e99 in flb_output_thread /home/don/src-ag/platform/fluent-bit/include/fluent-bit/flb_output.h:353
    #3 0x55cbd16d6e99 in tasks_start /home/don/src-ag/platform/fluent-bit/src/flb_engine_dispatch.c:92
    #4 0x55cbd16d7757 in flb_engine_dispatch /home/don/src-ag/platform/fluent-bit/src/flb_engine_dispatch.c:183
    #5 0x55cbd16d28ca in flb_engine_flush /home/don/src-ag/platform/fluent-bit/src/flb_engine.c:88
    #6 0x55cbd16d50a4 in flb_engine_handle_event /home/don/src-ag/platform/fluent-bit/src/flb_engine.c:271
    #7 0x55cbd16d50a4 in flb_engine_start /home/don/src-ag/platform/fluent-bit/src/flb_engine.c:515
    #8 0x55cbd16bc92d in flb_lib_worker /home/don/src-ag/platform/fluent-bit/src/flb_lib.c:391
    #9 0x7f1de05e66da in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76da)

Thread T1 created by T0 here:
    #0 0x7f1de0835d2f in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x37d2f)
    #1 0x55cbd19f1f72 in mk_utils_worker_spawn /home/don/src-ag/platform/fluent-bit/lib/monkey/mk_core/mk_utils.c:244
    #2 0x55cbd16bcb4f in flb_start /home/don/src-ag/platform/fluent-bit/src/flb_lib.c:419
    #3 0x55cbd16babce in flb_test_in_cpu_flush_2s_2times /home/don/src-ag/platform/fluent-bit/tests/runtime/in_cpu.c:96
    #4 0x55cbd16ba137 in test_do_run__ /home/don/src-ag/platform/fluent-bit/tests/runtime/../lib/acutest/acutest.h:464
    #5 0x55cbd16ba137 in test_run__ /home/don/src-ag/platform/fluent-bit/tests/runtime/../lib/acutest/acutest.h:546
    #6 0x55cbd16b8b7f in main /home/don/src-ag/platform/fluent-bit/tests/runtime/../lib/acutest/acutest.h:815
    #7 0x7f1ddf5b4b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

SUMMARY: AddressSanitizer: stack-buffer-overflow (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x79732) 
Shadow bytes around the buggy address:
  0x0c567fff8c10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8c20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8c30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8c40: 00 00 00 00 f1 f1 f1 f1 00 f2 f2 f2 f2 f2 f2 f2
  0x0c567fff8c50: 00 f2 f2 f2 f2 f2 f2 f2 00 00 f2 f2 f2 f2 f2 f2
=>0x0c567fff8c60: 00 00 00 00[f3]f3 f3 f3 00 00 00 00 00 00 00 00
  0x0c567fff8c70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c567fff8c80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8c90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8ca0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c567fff8cb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
                          struct flb_config *config)
  Right alloca redzone:    cb
==3558==ABORTING
FAILED: 1 of 1 unit tests have failed.

```

**Comment by edsiper** [comment]
thanks for reporting this problem!, Indeed the routine to allocate and copy the data was totally broken. 

Fixed by ab947dde

**Comment by premeximietas** [comment]
i have written the conf file of fluent-bit.conf as 
[SERVICE]
    Flush            5
    Daemon           Off
    Log_Level        info

[INPUT]
    Name             tail
    Path             C:\projects\logging_fastapi\logs_generated.log.*
    Tag              application.log
    Refresh_Interval 1

[FILTER]
    Name             record_modifier
    Match            *
    Record           container_id ${HOSTNAME}

[OUTPUT]
    Name             s3
    Match            application.log  
    Bucket           specyfeedbackbatch
    Region           us-east-1
    S3_Key_Format    /%Y/%m/%d/%H/%Y-%m-%d-%H-%M-%S-application.log
    Total_File_Size  5M
    
   it is not sending the log from the local to the aws s3 can any one state me what can be the possible reason for that and the logs of fluent bit are 
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [fluent bit] version=2.2.2, commit=eeea396e88, pid=1
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [cmetrics] version=0.6.6
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [ctraces ] version=0.4.0
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [input:tail:tail.0] initializing
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [fstore] created root path /tmp/fluent-bit/s3/specyfeedbackbatch
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [output:s3:s3.0] Using upload size 5000000 bytes
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [output:s3:s3.0] total_file_size is less than 10 MB, will use PutObject API
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [sp] stream processor started
2024-02-28 12:48:39 [2024/02/28 07:18:39] [ info] [output:s3:s3.0] worker #0 started

what is the possible reason for the issue ? i want it to send the logs from local to the s3 every time when ever the file is changed.
